### PR TITLE
Fix for jenkins results link and better handling of console errors

### PR DESCRIPTION
**Pull request by kelliejos**
Fix for broken Jenkins results link on build summary page and fix to prevent builds from hanging when there is a network hiccup while streaming the pipeline console.

### CODE REVIEW

File: Tasks/JenkinsQueueJobV2/job.ts [hunk 1]
```diff
@@ -424,8 +424,17 @@ export class Job {
...
-            } else if (httpResponse.statusCode != 200) {
+            } else if (httpResponse.statusCode == 401) {
+                    Util.failReturnCode(httpResponse, 'Job progress tracking failed to read job progress');
+                    thisJob.queue.TaskOptions.captureConsole = false;
+                    thisJob.queue.TaskOptions.capturePipeline = false;
+                    thisJob.queue.TaskOptions.shouldFail = true;
+                    thisJob.queue.TaskOptions.failureMsg = 'Job progress tracking failed to read job progress';
+                    thisJob.stopWork(0, JobState.Finishing);
+            } 
+            else if (httpResponse.statusCode != 200) {
```
Comment (brcrista) [inline Tasks/JenkinsQueueJobV2/job.ts]: style here seems to be `} else if (...) {`

Comment (brcrista) [inline Tasks/JenkinsQueueJobV2/job.ts]: Even though this file is using `==` in some places, it isn't type safe and I don't think it [should ever be used](https://stackoverflow.com/a/359509/3084634).  Can we fix at least this method to use use `===` and `!==`?

File: Tasks/JenkinsQueueJobV2/jenkinsqueuejobtask.ts [hunk 2]
```diff
@@ -92,6 +95,18 @@ export class TaskOptions {
 
         this.NO_CRUMB = 'NO_CRUMB';
         this.crumb = this.NO_CRUMB;
+
+        this.shouldFail = false;
+        this.failureMsg = '';
+    }
+}
+
+export class taskShouldFail {
```
Comment (brcrista) [inline Tasks/JenkinsQueueJobV2/jenkinsqueuejobtask.ts]: Is this used anywhere?

File: Tasks/JenkinsQueueJobV2/job.ts [hunk 3]
```diff
@@ -424,8 +424,17 @@ export class Job {
             } else if (httpResponse.statusCode == 404) {
                 // got here too fast, stream not yet available, try again in the future
                 thisJob.stopWork(thisJob.queue.TaskOptions.pollIntervalMillis, thisJob.State);
-            } else if (httpResponse.statusCode != 200) {
+            } else if (httpResponse.statusCode == 401) {
+                    Util.failReturnCode(httpResponse, 'Job progress tracking failed to read job progress');
```
Comment (brcrista) [inline Tasks/JenkinsQueueJobV2/job.ts]: Does this message need to be localized?

File: Tasks/JenkinsQueueJobV2/jenkinsqueuejobtask.ts (hunk 2 above)
Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/jenkinsqueuejobtask.ts]: Ah...no it's not. I had tried a different way of failing the task and this is a remnant of it.  I missed removing it when I ripped out the rest.

File: Tasks/JenkinsQueueJobV2/job.ts (hunk 1 above)
Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/job.ts]: I am fine with that. I usually use the !== and === forms, but was following the pattern of the file in this case. I'll update it.

File: Tasks/JenkinsQueueJobV2/job.ts [hunk 4]
```diff
@@ -437,7 +446,10 @@ export class Job {
                     thisJob.stopWork(0, JobState.Finishing);
                 }
             }
-        }).auth(thisJob.queue.TaskOptions.username, thisJob.queue.TaskOptions.password, true);
+        }).auth(thisJob.queue.TaskOptions.username, thisJob.queue.TaskOptions.password, true)
+        .on('error', (err) => { 
+            throw err;
```
Comment (lkillgore) [inline Tasks/JenkinsQueueJobV2/job.ts]: I think that `throw new Error(err)` is a better pattern.  It will preserve the stacktrace.

Comment (brcrista) [reply Tasks/JenkinsQueueJobV2/job.ts]: Shouldn't `err` already be an instance of the `Error` class?

Comment (lkillgore) [reply Tasks/JenkinsQueueJobV2/job.ts]: It probably is, but is it guaranteed to be?  Also, what do you want the stacktrace to look like?  I would want the top to be this line, and not something internal.

File: Tasks/JenkinsQueueJobV2/job.ts (hunk 3 above)
Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/job.ts]: The others were not, but I'll add this to the localized strings. They should probably be localized as well.

File: Tasks/JenkinsQueueJobV2/jobqueue.ts [hunk 5]
```diff
@@ -187,24 +187,24 @@ export class JobQueue {
...
             // if this job was joined to another follow that one instead
             job = findWorkingJob(job);
 
             if (job.ExecutableNumber === -1) {
-                jobContents += indent + job.Name + ' ' + colorize(job.GetResultString()) + '<br />\n';
+                jobContents += indent + job.Name + ' ' + colorize(job.GetResultString()) + '<br />';
             } else {
                 const url = job.ExecutableUrl && job.ExecutableUrl.replace('"', '%22');
-                jobContents += indent + '<a href="' + url + '">' + job.Name + ' #' + job.ExecutableNumber + '</a> ' + colorize(job.GetResultString()) + '<br />\n';
+                jobContents += indent + '[' + job.Name + ' #' + job.ExecutableNumber + '](' + url + ') ' + colorize(job.GetResultString()) + '<br />';
```
Comment (pjquirk) [inline Tasks/JenkinsQueueJobV2/jobqueue.ts]: This line has Markdown and a `<br />` tag, which doesn't seem right to me.  Is this method supposed to return a Markdown string or an HTML string?

Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/jobqueue.ts]: The Jenkins results markdown file is a mix of markdown and html. That is what was causing the links to not show up properly.  When the link is embedded in html, it is not rendered correctly in the UI, however we need the html to create the colored success/failure badge after the link.

File: Tasks/JenkinsQueueJobV2/job.ts (hunk 4 above)
Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/job.ts]: I concur, converting callbacks to promises is something that we need to do to properly "fix" the task's error handling.  However in the interim, this fix will prevent builds from hanging when we lose the network connection with Jenkins while streaming the pipeline console output.  I have a story in the backlog to update the task to use promises.

File: Tasks/JenkinsQueueJobV2/jobqueue.ts (hunk 5 above)
Comment (brcrista) [reply Tasks/JenkinsQueueJobV2/jobqueue.ts]: The `<br />` should be fine -- all HTML is also valid Markdown.

Comment (kelliejos) [reply Tasks/JenkinsQueueJobV2/jobqueue.ts]: Well...that depends on the markdown parser that is being used.  The one we use for the build summary page doesn't like mixes of markdown and html, or at least not markdown embedded in html. :)

**Pull request merged**